src/main/java/org/opensearch/knn/quantization/models/quantizationState/QuantizationStateCache.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.quantization.models.quantizationState;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.RemovalCause;
+import com.google.common.cache.RemovalNotification;
+import lombok.extern.log4j.Log4j2;
+import org.opensearch.cluster.service.ClusterService;
+
+import java.io.IOException;
+import java.time.Instant;
+import java.util.concurrent.TimeUnit;
+
+import static org.opensearch.knn.index.KNNSettings.QUANTIZATION_STATE_CACHE_EXPIRY_TIME_MINUTES_SETTING;
+import static org.opensearch.knn.index.KNNSettings.QUANTIZATION_STATE_CACHE_SIZE_LIMIT_SETTING;
+
+/**
+ * A thread-safe singleton cache that contains quantization states.
+ */
+@Log4j2
+public class QuantizationStateCache {
+
+    private static volatile QuantizationStateCache instance;
+    private static ClusterService clusterService;
+    private Cache<String, QuantizationState> cache;
+    private long maxCacheSizeInKB;
+    private Instant evictedDueToSizeAt;
```

Comment:
shouldn't this just be a counter of type long?

Reply:
Not sure I understand, why would it be a counter instead of just a long holding the max size (which is calculated from the percent setting)